File: _includes/code/howto/pq-compression.py
# THIS FILE NEEDS TESTS

# ==============================
# =====  DOWNLOAD DATA =====
# ==============================

# START DownloadData
import requests
import json

# Download the data
resp = requests.get(
    "https://raw.githubusercontent.com/weaviate-tutorials/intro-workshop/main/data/jeopardy_1k.json"
)

# Load the data so you can see what it is
data = json.loads(resp.text)

# Parse the JSON and preview it
print(type(data), len(data))
print(json.dumps(data[1], indent=2))

# END DownloadData

# ==============================
# =====  CONNECT =====
# ==============================

# START ConnectCode
import weaviate, os, json
import weaviate.classes as wvc

client = weaviate.connect_to_local(
    headers={
        "X-OpenAI-Api-Key": os.environ[
            "OPENAI_API_KEY"
        ]  # Replace with your OpenAI API key
    }
)

client.is_ready()

# END ConnectCode

# ==============================
# =====  INITIAL SCHEMA =====
# ==============================

if (client.collections.exists("Question")):
    client.collections.delete("Question")

# START InitialSchema
client.collections.create(
    name="Question",
    description="A Jeopardy! question",
    vectorizer_config=wvc.Configure.Vectorizer.text2vec_openai(),
    generative_config=wvc.Configure.Generative.openai(),
    properties=[
        wvc.Property(name="title", data_type=wvc.DataType.TEXT),
    ],
)

# END InitialSchema

# ==============================
# =====  LOAD DATA =====
# ==============================

# START LoadData
def parse_data():
    object_list = []
    for obj in data:
        object_list.append(
            {
                "question": obj["Question"],
                "answer": obj["Answer"],
                "round": obj["Round"],
            }
        )

    return object_list

jeopardy = client.collections.get("Question")
jeopardy.data.insert_many(parse_data())

# Check upload
response = jeopardy.aggregate.over_all(total_count=True)

# Should equal the number of objects uploaded
print(response.total_count)

# END LoadData

# ==============================
# =====  UPDATE SCHEMA =====
# ==============================

# START UpdateSchema
import weaviate.classes as wvc

jeopardy = client.collections.get("Question")
jeopardy.config.update(
    vector_index_config=wvc.Reconfigure.VectorIndex.hnsw(
        quantizer=wvc.Reconfigure.VectorIndex.Quantizer.pq()
    )
)


# END UpdateSchema

# ==============================
# =====  GET THE SCHEMA =====
# ==============================

# START GetSchema
jeopardy = client.collections.get("Question")
config = jeopardy.config.get()
pq_config = config.vector_index_config.pq

# print some of the config properties
print(f"Enabled: { pq_config.enabled }")
print(f"Training: { pq_config.training_limit }")
print(f"Segments: { pq_config.segments }")
print(f"Centroids: { pq_config.centroids }")

# END GetSchema

File: _includes/code/howto/pq-compression.go
package main

import (
	"context"
	"encoding/json"
	"log"
	"net/http"
	"os"

	"github.com/weaviate/weaviate-go-client/v4/weaviate"
	"github.com/weaviate/weaviate/entities/models"
)

const url = "https://raw.githubusercontent.com/weaviate-tutorials/intro-workshop/main/data/jeopardy_1k.json"

func main() {
	// START DownloadData
	resp, err := http.Get(url)
	if err != nil {
		 log.Fatalf("get dataset: %v", err)
	}

	var data []struct {
			Question string `json:"Question"`
			Answer   string `json:"Answer"`
	}
 
	err = json.NewDecoder(resp.Body).Decode(&data)
	if err != nil {
		 log.Fatalf("read dataset: %v", err)
	}

	log.Printf("data: %+v", data)
	// END DownloadData

	// START ConnectCode
	client, err := weaviate.NewClient(weaviate.Config{
			Scheme: "http",
			Host:   "localhost:8080",
			Headers: map[string]string{
			 	"X-OpenAI-Api-Key": os.Getenv("OPENAI_APIKEY"),
			},
	})
	
	if err != nil {
	 	log.Fatalf("init client: %v", err)
	}
	// END ConnectCode

	// START InitialSchema
	class := &models.Class{
			Class:      "Question",
			Vectorizer: "text2vec-openai",
			Properties: []*models.Property{
					{Name: "Question", DataType: []string{"text"}},
					{Name: "Answer", DataType: []string{"text"}},
			},
	}
	
	err = client.Schema().ClassCreator().
		 WithClass(class).Do(context.Background())
	
	if err != nil {
	 	log.Fatalf("create class: %v", err)
	}
	// END InitialSchema

	// START LoadData
	batcher := client.Batch().ObjectsBatcher()
	for _, d := range data {
			batcher.WithObjects(&models.Object{
					Class: "Question",
					Properties: map[string]interface{}{
							"question": d.Question,
							"answer":   d.Answer,
					},
			})
	}
	
	batch, err := batcher.Do(context.Background())
	if err != nil {
	 	log.Fatalf("batcher: %v", err)
	}
	
	for i, b := range batch {
			if b.Result.Errors != nil {
					for _, err := range b.Result.Errors.Error {
						 log.Print(err.Message)
				 	}
					log.Fatalf("object@idx %d had batch errors above", i)
			}
	}
	// END LoadData

	// START UpdateSchema
	class, err = client.Schema().ClassGetter().
	 	WithClassName("Question").Do(context.Background())
	
	if err != nil {
	 	log.Fatalf("get class for vec idx cfg update: %v", err)
	}

	cfg := class.VectorIndexConfig.(map[string]interface{})
	cfg["pq"] = map[string]interface{}{
			"enabled":       true,
			"trainingLimit": 100_000,
			"segments":      96,
	}
	class.VectorIndexConfig = cfg
	
	err = client.Schema().ClassUpdater().
		 WithClass(class).Do(context.Background())
	
	if err != nil {
	 	log.Fatalf("update class to use pq: %v", err)
	}
	// END UpdateSchema

	// START GetSchema
	class, err = client.Schema().ClassGetter().
	 	WithClassName("Question").Do(context.Background())
	if err != nil {
	 	log.Fatalf("get class to verify vec idx cfg changes: %v", err)
	}
	
	cfg = class.VectorIndexConfig.(map[string]interface{})
	log.Printf("pq config: %v", cfg["pq"])
	// END GetSchema
}
